File: P1573 Number of Ways to Split a String/solution.py
class Solution:
    def numWays(self, s: str) -> int:
        mod_num = 10 ** 9 + 7
        num_of_ones = s.count('1')
        if num_of_ones % 3 != 0:
            return 0
        elif num_of_ones == 0:
            return (len(s) - 1) * (len(s) - 2) // 2 % mod_num
        else:
            ones_in_each_part = num_of_ones / 3
            ways_before_first_cut = 0
            ways_between_first_and_second_cut = 0
            count_of_one = 0
            for c in s:
                if c == '1':
                    count_of_one += 1
                if count_of_one == ones_in_each_part:
                    ways_before_first_cut +=1
                elif count_of_one == 2 * ones_in_each_part:
                    ways_between_first_and_second_cut += 1
            return ways_before_first_cut * ways_between_first_and_second_cut % mod_num

File: P1573 Number of Ways to Split a String/Solution.java
class Solution {
    public int numWays(String s) {
        int modNum = 1000000007;
        int numOfOnes = countOnes(s);
        if (numOfOnes % 3 != 0) {
            return 0;
        } else if(numOfOnes == 0) {
            return (int) ((s.length() - 1L) * (s.length() - 2L) / 2 % modNum);
        } else {
            int onesInEachPart =  numOfOnes / 3;
            long waysBeforeFirstCut = 0, waysBetweenFirstAndSecondCut = 0;
            int count = 0;
            for (int i = 0; i < s.length(); i++) {
                if (s.charAt(i) == '1') {
                    count++;
                }
                if (count == onesInEachPart) { 
                    waysBeforeFirstCut++;
                } else if (count == onesInEachPart * 2) {
                    waysBetweenFirstAndSecondCut++;
                }
            }
            return (int) (waysBeforeFirstCut * waysBetweenFirstAndSecondCut % modNum);
        }
    }
    
    private static int countOnes(String s) {
        int count = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '1') {
                count++;
            }
        }
        return count;
    }
}